File: debot/src/main/java/com/tomoima/debot/Debot.java
package com.tomoima.debot;


import android.content.Context;
import android.hardware.SensorManager;
import android.support.v4.app.FragmentActivity;

import com.squareup.seismic.ShakeDetector;

import java.lang.ref.WeakReference;

public class Debot {
    private DebotDialog debotDialog;
    private ShakeDetector sd;
    private SensorManager sensorManager;
    private WeakReference<FragmentActivity> activityWeakRef;
    private boolean canShake;

    private static Debot instance = new Debot();

    private Debot(){
        debotDialog = DebotDialog.getInstance();
    }

    public static Debot getInstance() {
        return instance;
    }

    public void allowShake(Context context) {
        canShake = true;
        setupSensor(context);
    }

    public void startSensor(FragmentActivity activity) {
        if(!canShake) return;
        activityWeakRef = new WeakReference<>(activity);
        if(sd != null) {
            sd.start(sensorManager);
        }
    }

    public void stopSensor() {
        if (!canShake) return;
        if (activityWeakRef != null) {
            activityWeakRef.clear();
        }
        if (sd != null) {
            sd.stop();
        }
    }

    public void showDebugMenu(FragmentActivity activity) {
        debotDialog.showDebugMenu(activity);
    }

    private void setupSensor(Context context) {
        sensorManager = (SensorManager) context.getSystemService(Context.SENSOR_SERVICE);
        sd = new ShakeDetector(new ShakeDetector.Listener() {
            @Override
            public void hearShake() {
                if(activityWeakRef.get() != null) {
                    showDebugMenu(activityWeakRef.get());
                }
            }
        });
        sd.setSensitivity(ShakeDetector.SENSITIVITY_LIGHT);
    }

}

File: debot/src/main/java/com/tomoima/debot/Debot.kt
package com.tomoima.debot


import android.content.Context
import android.hardware.SensorManager
import androidx.fragment.app.FragmentActivity

import com.squareup.seismic.ShakeDetector

import java.lang.ref.WeakReference

class Debot private constructor() {
    private var sd: ShakeDetector? = null
    private var sensorManager: SensorManager? = null
    private var activityWeakRef: WeakReference<FragmentActivity>? = null
    private var canShake: Boolean = false

    fun allowShake(context: Context) {
        canShake = true
        setupSensor(context)
    }

    @JvmOverloads
    fun startSensor(activity: FragmentActivity, sensorDelay: Int = SensorManager.SENSOR_DELAY_GAME) {
        if (!canShake) return
        activityWeakRef = WeakReference(activity)
        sd?.start(sensorManager, sensorDelay)
    }

    fun stopSensor() {
        if (!canShake) return
        activityWeakRef?.clear()
        sd?.stop()
    }

    fun showDebugMenu(activity: FragmentActivity) {
        DebotDialog.get().showDebugMenu(activity)
    }

    private fun setupSensor(context: Context) {
        sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
        sd = ShakeDetector(ShakeDetector.Listener {
            activityWeakRef?.get()?.let(::showDebugMenu)
        })
        sd?.setSensitivity(ShakeDetector.SENSITIVITY_LIGHT)
    }

    companion object {
        @JvmStatic fun getInstance(): Debot = Debot()
    }

}
